### PR TITLE
Add function metadata

### PR DESCRIPTION
**Pull request by dain**
Use function metadata for deterministic checks

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/sql/planner/DeterminismEvaluator.java [hunk 1]
```diff
@@ -33,33 +31,45 @@
...
         AtomicBoolean deterministic = new AtomicBoolean(true);
-        new Visitor().process(expression, deterministic);
+        new Visitor(isDeterministic).process(expression, deterministic);
         return deterministic.get();
     }
 
     private static class Visitor
             extends DefaultExpressionTraversalVisitor<Void, AtomicBoolean>
     {
-        private static final Set<String> FUNCTIONS = ImmutableSet.of("rand", "random", "shuffle", "uuid");
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/sql/planner/DeterminismEvaluator.java]: ❤️

File: presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java [hunk 2]
```diff
@@ -2203,12 +2205,19 @@ private void verifySelectDistinct(QuerySpecification node, List<Expression> outp
                     throw semanticException(EXPRESSION_NOT_IN_DISTINCT, node.getSelect(), "For SELECT DISTINCT, ORDER BY expressions must appear in select list");
                 }
 
-                if (!isDeterministic(expression)) {
+                if (!DeterminismEvaluator.isDeterministic(this::isDeterministic, expression)) {
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java]: This looks strange. I'd change the `DeterminismEvaluator.isDeterministic` method to take a `Function<FunctionCall, FunctionMetadata>` and leave it up to DeterminismEvaluator to call `FunctionMetadata.isDeterministic()`. This callsite would then be:

```java
if (!DeterminismEvaluator.isDeterministic(expression, this::getFunctionMetadata)) {
   ...
}
```

And the function below:

```java
private FunctionMetadata getFunctionMetadata(FunctionCall functionCall)
{
    ResolvedFunction resolvedFunction = analysis.getResolvedFunction(functionCall);
    verify(resolvedFunction != null, "function has not been analyzed yet: %s", functionCall);
    return metadata.getFunctionMetadata(resolvedFunction);
}
```

File: presto-main/src/main/java/io/prestosql/metadata/FunctionMetadata.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+import static java.util.Objects.requireNonNull;
+
+public class FunctionMetadata
+{
+    private final FunctionId functionId;
+    private final Signature signature;
+    private final boolean hidden;
+    private final boolean deterministic;
+    private final String description;
```
[the remark this answers is not in the document]
Comment (dain) [reply presto-main/src/main/java/io/prestosql/metadata/FunctionMetadata.java]: This is `requireNonNull`

File: presto-main/src/main/java/io/prestosql/metadata/SqlOperator.java [hunk 4]
```diff
@@ -23,35 +23,25 @@
...
 {
-    protected SqlOperator(OperatorType operatorType, List<TypeVariableConstraint> typeVariableConstraints, List<LongVariableConstraint> longVariableConstraints, TypeSignature returnType, List<TypeSignature> argumentTypes)
+    protected SqlOperator(
+            OperatorType operatorType,
+            List<TypeVariableConstraint> typeVariableConstraints,
+            List<LongVariableConstraint> longVariableConstraints,
+            TypeSignature returnType,
+            List<TypeSignature> argumentTypes)
     {
         // TODO This should take Signature!
```
[the remark this answers is not in the document]
Comment (dain) [reply presto-main/src/main/java/io/prestosql/metadata/SqlOperator.java]: I think the comment is the `SqlOperator` constructor should just take a signature,

File: presto-main/src/main/java/io/prestosql/metadata/FunctionMetadata.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (martint) [reply presto-main/src/main/java/io/prestosql/metadata/FunctionMetadata.java]: I was thinking we might want to not force missing descriptions to become “”

[the remark this answers is not in the document]
Comment (martint) [reply presto-main/src/main/java/io/prestosql/metadata/FunctionMetadata.java]: But no strong preference. Not sure it matters

[the remark this answers is not in the document]
Comment (findepi) [reply presto-main/src/main/java/io/prestosql/metadata/FunctionMetadata.java]: would you disallow `""` as a description? i guess no, so using is by default should not harm either.

[the remark this answers is not in the document]
Comment (dain) [reply presto-main/src/main/java/io/prestosql/metadata/FunctionMetadata.java]: Empty string is used for operators. I think everything else has a description.

**Pull request merged**